File: backend/app/api/v1/agents.py
"""Agent system API endpoints — trust, approvals, email config, activity."""

from uuid import UUID

from fastapi import APIRouter, Depends, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.core.database import get_db
from app.models.approval_request import ApprovalRequest, ApprovalStatus
from app.models.user import User
from app.schemas.agent import (
    TrustOverview,
    TrustScoreResponse,
    TrustUpdateRequest,
    ApprovalDecision,
    ApprovalRequestResponse,
    EmailConfigCreate,
    EmailConfigResponse,
)
from app.services.trust import TrustService

router = APIRouter(tags=["Agents"])


@router.get("/trust", response_model=TrustOverview)
async def get_trust_scores(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get all trust scores for the current user."""
    service = TrustService(db)
    scores = await service.get_all_scores(str(current_user.id))
    return TrustOverview(
        scores=[TrustScoreResponse.model_validate(s) for s in scores]
    )


@router.put("/trust", status_code=status.HTTP_200_OK)
async def set_trust_level(
    data: TrustUpdateRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Set trust level for an agent type (applies to all action types)."""
    service = TrustService(db)
    for action_type in ["read", "write", "send", "delete"]:
        await service.set_trust_level(str(current_user.id), data.agent_type, action_type, data.trust_level)
    await db.commit()
    return {"status": "updated", "agent_type": data.agent_type, "trust_level": data.trust_level}


@router.get("/approvals/pending", response_model=list[ApprovalRequestResponse])
async def get_pending_approvals(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get all pending approval requests for the current user."""
    stmt = select(ApprovalRequest).where(
        ApprovalRequest.user_id == current_user.id,
        ApprovalRequest.status == ApprovalStatus.PENDING,
    ).order_by(ApprovalRequest.created_at.desc())
    result = await db.execute(stmt)
    return [ApprovalRequestResponse.model_validate(r) for r in result.scalars().all()]


@router.post("/approve/{approval_id}")
async def approve_action(
    approval_id: UUID,
    decision: ApprovalDecision,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Approve or reject a pending action."""
    stmt = select(ApprovalRequest).where(
        ApprovalRequest.id == approval_id,
        ApprovalRequest.user_id == current_user.id,
        ApprovalRequest.status == ApprovalStatus.PENDING,
    )
    result = await db.execute(stmt)
    request = result.scalar_one_or_none()
    if not request:
        return {"error": "Approval request not found or already resolved"}

    request.status = ApprovalStatus.APPROVED if decision.approved else ApprovalStatus.REJECTED
    request.user_reason = decision.reason
    await db.commit()

    trust_service = TrustService(db)
    if decision.approved:
        await trust_service.record_action(str(current_user.id), request.agent_type, request.action, success=True)
    else:
        await trust_service.record_violation(str(current_user.id), request.agent_type, request.action)
    await db.commit()

    return {"status": request.status.value, "approval_id": str(approval_id)}


@router.post("/email/config", response_model=EmailConfigResponse)
async def save_email_config(
    data: EmailConfigCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Save email configuration (SMTP/IMAP)."""
    from app.services.email import EmailService
    service = EmailService(db)
    config = await service.save_config(
        user_id=str(current_user.id),
        smtp_host=data.smtp_host,
        smtp_port=data.smtp_port,
        smtp_user=data.smtp_user,
        smtp_password=data.smtp_password,
        imap_host=data.imap_host,
        imap_port=data.imap_port,
        imap_user=data.imap_user,
        imap_password=data.imap_password,
    )
    await db.commit()
    return EmailConfigResponse.model_validate(config)


@router.get("/email/config", response_model=EmailConfigResponse | None)
async def get_email_config(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get email configuration for the current user."""
    from app.services.email import EmailService
    service = EmailService(db)
    config = await service.get_config(str(current_user.id))
    if not config:
        return None
    return EmailConfigResponse.model_validate(config)

File: mobile/services/agents.ts
import api from "./api";

export interface TrustScore {
  agent_type: string;
  action_type: string;
  trust_level: number;
  successful_actions: number;
  total_actions: number;
}

export interface TrustOverview {
  scores: TrustScore[];
}

export interface ApprovalRequest {
  id: string;
  agent_type: string;
  action: string;
  action_details: Record<string, unknown>;
  status: string;
  timeout_seconds: number;
  expires_at: string | null;
  created_at: string;
}

export interface EmailConfig {
  id: string;
  smtp_host: string;
  smtp_port: number;
  smtp_user: string;
  imap_host: string;
  imap_port: number;
  imap_user: string;
  is_active: boolean;
}

export interface EmailConfigCreate {
  smtp_host: string;
  smtp_port: number;
  smtp_user: string;
  smtp_password: string;
  imap_host: string;
  imap_port: number;
  imap_user: string;
  imap_password: string;
}

export interface AgentActivity {
  id: string;
  agent_type: string;
  action: string;
  status: string;
  details: Record<string, unknown> | null;
  result: string | null;
  duration_ms: number | null;
  error: string | null;
  created_at: string;
}

export const agentApi = {
  getTrustScores: async (): Promise<TrustOverview> => {
    const response = await api.get<TrustOverview>("/agents/trust");
    return response.data;
  },

  setTrustLevel: async (agent_type: string, trust_level: number) => {
    const response = await api.put("/agents/trust", { agent_type, trust_level });
    return response.data;
  },

  getPendingApprovals: async (): Promise<ApprovalRequest[]> => {
    const response = await api.get<ApprovalRequest[]>("/agents/approvals/pending");
    return response.data;
  },

  approve: async (approvalId: string, approved: boolean, reason?: string) => {
    const response = await api.post(`/agents/approve/${approvalId}`, { approved, reason });
    return response.data;
  },

  getEmailConfig: async (): Promise<EmailConfig | null> => {
    const response = await api.get<EmailConfig | null>("/agents/email/config");
    return response.data;
  },

  saveEmailConfig: async (config: EmailConfigCreate): Promise<EmailConfig> => {
    const response = await api.post<EmailConfig>("/agents/email/config", config);
    return response.data;
  },
};
